### ISSUE TITLE
Upsample Module exist bug

### ISSUE THREAD

**Issue by BBuf** [pytorch/pytorch#65200]
## 🐛 Bug

When using the Upsample Module, the results calculated by the CPU and GPU are different under certain settings. 

## To Reproduce

When I use the CPU, 

```python
import torch

x = torch.tensor([[[[0.4314, 0.0482, 0.4077, 0.5083, 0.0997, 0.0557, 0.6928],
          [0.9508, 0.5647, 0.9578, 0.7280, 0.3346, 0.6295, 0.7685],
          [0.9874, 0.6847, 0.2015, 0.5326, 0.2211, 0.2419, 0.6180],
          [0.7117, 0.0984, 0.9643, 0.4791, 0.5381, 0.0221, 0.3342]]]], dtype=torch.float32, requires_grad=True, device="cpu")

m = torch.nn.Upsample(scale_factor=1.0239041660002521, mode='bilinear', align_corners=False)
m = m.to("cpu")
y = m(x)
print(y.shape)
y = y.sum()
y.backward()
print(x.grad.cpu().numpy())
```

The result is:

torch.Size([1, 1, 4, 7])
[[[[1. 1. 1. 1. 1. 1. 1.]
   [1. 1. 1. 1. 1. 1. 1.]
   [1. 1. 1. 1. 1. 1. 1.]
   [1. 1. 1. 1. 1. 1. 1.]]]]

When I use gpu:

```python
import torch

x = torch.tensor([[[[0.4314, 0.0482, 0.4077, 0.5083, 0.0997, 0.0557, 0.6928],
          [0.9508, 0.5647, 0.9578, 0.7280, 0.3346, 0.6295, 0.7685],
          [0.9874, 0.6847, 0.2015, 0.5326, 0.2211, 0.2419, 0.6180],
          [0.7117, 0.0984, 0.9643, 0.4791, 0.5381, 0.0221, 0.3342]]]], dtype=torch.float32, requires_grad=True, device="cuda")

m = torch.nn.Upsample(scale_factor=1.0239041660002521, mode='bilinear', align_corners=False)
m = m.to("cuda")
y = m(x)
print(y.shape)
y = y.sum()
y.backward()
print(x.grad.cpu().numpy())

```

The result is：

torch.Size([1, 1, 4, 7])
[[[[1.0712646  1.059183   1.0591829  1.0591826  1.0591829  1.0591826
    0.8779554 ]
   [1.059183   1.0472374  1.0472374  1.0472372  1.0472374  1.0472372
    0.86805385]
   [1.0591829  1.0472374  1.0472374  1.0472372  1.0472374  1.0472372
    0.8680538 ]
   [0.9504461  0.93972695 0.939727   0.9397268  0.93972695 0.9397268
    0.77893853]]]]

## Pytorch Version

'1.9.0+cu102'

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @ngimel

**Comment by BBuf** [comment]
Report a bug again, when I run the following code：

```python
import torch

x = torch.tensor([[[[0.0990],
          [0.4117],
          [0.6456],
          [0.4370]]]], dtype=torch.float32, requires_grad=True)

m = torch.nn.Upsample(scale_factor=2.16, mode='nearest')
y = m(x)
y = y.sum()
y.backward()
print(x.grad.numpy())
```

I get the result:

```
[[[[4.]
   [4.]
   [4.]
   [4.]]]]
```

I think this gradient is wrong, because in Pytorch's coordinate mapping function, the gradient should go to the last branch in the following code. But because of rounding, the size of the output is twice the size of the input, so the second branch is actually taken in backward.

```c++
static inline int64_t nearest_idx(
    int64_t output_index,
    int64_t input_size,
    int64_t output_size,
    c10::optional<double> scales) {
  if (output_size == input_size) {
    // scale_factor = 1, simply copy
    return output_index;
  } else if (output_size == 2 * input_size) {
    // scale_factor = 2, shift input index
    return output_index >> 1;
  } else {
    float scale = compute_scales_value<float>(scales, input_size, output_size);
    return nearest_neighbor_compute_source_index(scale, output_index, input_size);
  }
}
```

In PaddlePaddle and OneFlow, We get the same gradient output ：

```
[[[[6.]
   [4.]
   [4.]
   [2.]]]]
```

If you can confirm this bug, I can open a pr to fix it.

**Comment by bdhirsh** [comment]
Marking hi-pri for the silent correctness bug

**Comment by bdhirsh** [comment]
I can repro it on 1.9.0. However, it looks like it's fixed on master - I can't repro it there.

**Comment by bdhirsh** [comment]
leaving as hi-pri to confirm the bug described in https://github.com/pytorch/pytorch/issues/65200#issuecomment-921535994 is still on master.

@malfet: would be nice to bisect the commit that fixed the first bug on this issue and patch into the release branch
